File: packages/study-logic/src/store.ts
import type { Attempt, Quiz, QuizItem, Topic } from "./types.js";

export class MemoryStore {
  private readonly topics = new Map<string, Topic[]>();
  private readonly quizzes = new Map<string, Quiz>();
  private readonly items = new Map<string, QuizItem>();
  private readonly attempts: Attempt[] = [];

  listTopics(notebookId: string): Topic[] {
    return (this.topics.get(notebookId) ?? []).map((t) => ({ ...t }));
  }

  setTopics(notebookId: string, topics: Topic[]): void {
    this.topics.set(
      notebookId,
      topics.map((t) => ({ ...t })),
    );
  }

  putQuiz(quiz: Quiz, items: QuizItem[]): void {
    this.quizzes.set(quiz.id, { ...quiz, item_ids: [...quiz.item_ids] });
    for (const item of items) {
      this.items.set(item.id, {
        ...item,
        topic_ids: [...item.topic_ids],
        citation_chunk_ids: [...item.citation_chunk_ids],
        choices: item.choices.map((c) => ({ ...c })),
      });
    }
  }

  getQuiz(quizId: string): Quiz | undefined {
    const quiz = this.quizzes.get(quizId);
    return quiz ? { ...quiz, item_ids: [...quiz.item_ids] } : undefined;
  }

  getItem(itemId: string): QuizItem | undefined {
    const item = this.items.get(itemId);
    if (!item) return undefined;
    return {
      ...item,
      topic_ids: [...item.topic_ids],
      citation_chunk_ids: [...item.citation_chunk_ids],
      choices: item.choices.map((c) => ({ ...c })),
    };
  }

  addAttempt(attempt: Attempt): void {
    this.attempts.push({ ...attempt, topic_ids: [...attempt.topic_ids] });
  }

  listAttemptsForNotebook(notebookId: string): Attempt[] {
    return this.attempts
      .filter((a) => a.notebook_id === notebookId)
      .map((a) => ({ ...a, topic_ids: [...a.topic_ids] }));
  }
}

File: packages/study_logic/src/study_logic/store.py
from __future__ import annotations

from study_logic.models import Attempt, Chat, ChatMessage, Handoff, Quiz, QuizItem, Topic


class MemoryStore:
    def __init__(self) -> None:
        self._topics: dict[str, list[Topic]] = {}
        self._quizzes: dict[str, Quiz] = {}
        self._items: dict[str, QuizItem] = {}
        self._attempts: list[Attempt] = []
        self._chats: dict[str, Chat] = {}
        self._messages: list[ChatMessage] = []
        self._handoffs: list[Handoff] = []

    def list_topics(self, notebook_id: str) -> list[Topic]:
        return list(self._topics.get(notebook_id, []))

    def set_topics(self, notebook_id: str, topics: list[Topic]) -> None:
        self._topics[notebook_id] = list(topics)

    def put_quiz(self, quiz: Quiz, items: list[QuizItem]) -> None:
        self._quizzes[quiz.id] = quiz
        for item in items:
            self._items[item.id] = item

    def get_quiz(self, quiz_id: str) -> Quiz | None:
        return self._quizzes.get(quiz_id)

    def get_item(self, item_id: str) -> QuizItem | None:
        return self._items.get(item_id)

    def add_attempt(self, attempt: Attempt) -> None:
        self._attempts.append(attempt)

    def list_attempts(self, notebook_id: str) -> list[Attempt]:
        return [a for a in self._attempts if a.notebook_id == notebook_id]

    def put_chat(self, chat: Chat) -> None:
        self._chats[chat.id] = chat

    def get_chat(self, chat_id: str) -> Chat | None:
        return self._chats.get(chat_id)

    def list_chats(self, notebook_id: str) -> list[Chat]:
        chats = [chat for chat in self._chats.values() if chat.notebook_id == notebook_id]
        chats.sort(key=lambda chat: (0 if chat.kind == "orchestrator" else 1, chat.created_at, chat.id))
        return chats

    def get_orchestrator(self, notebook_id: str) -> Chat | None:
        for chat in self.list_chats(notebook_id):
            if chat.kind == "orchestrator":
                return chat
        return None

    def list_open_specialists(self, notebook_id: str) -> list[Chat]:
        return [chat for chat in self.list_chats(notebook_id) if chat.kind == "specialist" and chat.status == "open"]

    def add_message(self, message: ChatMessage) -> None:
        self._messages.append(message)

    def list_messages(self, chat_id: str) -> list[ChatMessage]:
        return [message for message in self._messages if message.chat_id == chat_id]

    def add_handoff(self, handoff: Handoff) -> None:
        self._handoffs.append(handoff)

    def list_handoffs(self, notebook_id: str) -> list[Handoff]:
        chat_ids = {chat.id for chat in self.list_chats(notebook_id)}
        rows = [row for row in self._handoffs if row.from_chat_id in chat_ids or row.to_chat_id in chat_ids]
        rows.sort(key=lambda row: row.created_at)
        return rows

    def get_handoff_for_chat(self, from_chat_id: str) -> Handoff | None:
        for row in self._handoffs:
            if row.from_chat_id == from_chat_id:
                return row
        return None
